client/lib/plans/features-list.tsx

```diff
@@ -1769,6 +1780,22 @@ export const FEATURES_LIST: FeatureList = {
 		getAlternativeTitle: () => '0%',
 		getFeatureGroup: () => FEATURE_GROUP_PAYMENT_TRANSACTION_FEES,
 	},
+	[ FEATURE_50GB_STORAGE_ADD_ON ]: {
```

Comment:
To top it up, I would also question the need for `FEATURE_50GB_STORAGE_ADD_ON` as a separate object here. It's just a repeated structure to `FEATURE_50GB_STORAGE` for the most part with a different slug right? Obviously, alternative (to use same structure) needs a little deeper consideration.

Reply:
The distinct difference is the inclusion of `getQuantity` on the feature object. https://github.com/Automattic/wp-calypso/pull/79041/files/afe13155242a279e041c346515508fe2383fd22a#r1279555091

Reply:
Hmm gotcha. Potentially separate things. Although, both could be described in terms of `x * unit` right? 🤔

Reply:
>  Although, both could be described in terms of x * unit right? 🤔

We could 👍  My decision to separate was more of an intuition -- it seems natural to see them as two different concepts, and I wouldn't be surprised if there's more divergent attributes in the future, but to answer your question, yes. As of right now, they both could be described in terms of x * unit.

Reply:
> If we feel strongly about joining the two, I can make that happen.

I don't feel too strongly. On one end, I feel a separate object allows us to "do more". On the other end, I feel strongly about NOT replicating anything, and YAGNI (are we actually going to have any unique functionality for these "add-on" features)? 

Let's maybe confirm whether we have standard features being used as add-ons already. I think I'm seeing a few in `my-sites/add-ons/hooks/use-add-ons`. If that's the case, then we have a mixture of things happening already:

1. new features created that are prefixed with `add_on` and for the most part replicate the original "non add-on" feature to introduce other wording, properties, etc.
2. feature definitions used as both add-ons and plan features (so no `add_on` prefix, probably no alternative titles).

Make the best judgment. with @southp pushing for this to move forward, I'll retract from this conversation.

Reply:
>  I think I'm seeing a few in my-sites/add-ons/hooks/use-add-ons

Ah I think I fully grasp what you're saying now. Thanks for the explanation. I'll think this over 🤔

Reply:
> In that situation it seems like the natural next step is to duplicate use-add-on into client/my-sites/plan-features-2023-grid/hooks

I started the process in a preliminary PR here https://github.com/Automattic/wp-calypso/pull/80114 so we can continue discussion in a PR with fewer historical conversations in it

Reply:
Thanks for starting on a follow-up PR already, @jeyip .

I haven't understood the whole picture enough to validate the direction of https://github.com/Automattic/wp-calypso/pull/80114 yet, but I have a feeling that what we want to resolve here should be a data definition level thing, which hasn't got to the level of APIs (e.g. the hooks) yet.

At the conceptual level, a product contains a feature set. Just like a plan provides a defined set of features, an add-on should also provide a defined set of features. Thus, plans, add-ons, and features should be separated concept. Thus, intuitively speaking, the storage add-ons should be able to reuse the existing features. For example, FEATURE_50GB_STORAGE should be what is provided by the 50G add-on.

However, I think we can't do that now since the existing FEATURE_50GB_STORAGE and what's provided by these add-ons are actually different:

The storage features that are used by the plans right now are simple: e.g. FEATURE_50GB_STORAGE is just one feature that grants 50G storage.
FEATURE_50GB_STORAGE_ADD_ON is different: it's 50 accumulated "1GB storage" feature provided by PRODUCT_1GB_SPACE product.
Thus, these two are not interchangeable in the current form. We can ask whether the client side needs to know this much implementation details(I guess we need that for the checkout? Otherwise we can just reuse the same storage feature), or come up with a better data structure and API to represent this concept. Before we can do any of them, they'd be better separated with some inline explanation in the code.

Reply:
> At the conceptual level, a product contains a feature set. Just like a plan provides a defined set of features, an add-on should also provide a defined set of features. Thus, plans, add-ons, and features should be separated concept. Thus, intuitively speaking, the storage add-ons should be able to reuse the existing features. For example, FEATURE_50GB_STORAGE should be what is provided by the 50G add-on.

Interesting. This conceptual breakdown is helpful

> I guess we need that for the checkout?

Yes you're totally right 👍 

> or come up with a better data structure and API to represent this concept.

If I understand Christos correctly, we have two ways of defining storage add ons right now. 

One set of storage add on definitions exists here in the `features-list`. The other set exists in the `use-add-on` hook. 

https://github.com/Automattic/wp-calypso/blob/4581bd60b113213f90cc65119c75699434132bda/client/my-sites/add-ons/hooks/use-add-ons.ts#L78-L101

In the follow-up I linked, I'm attempting to unify the definition that we use so that we have one source of truth. https://github.com/Automattic/wp-calypso/pull/80114

Reply:
> However, I think we can't do that now since the existing FEATURE_50GB_STORAGE and what's provided by these add-ons are actually different:
> 
> The storage features that are used by the plans right now are simple: e.g. FEATURE_50GB_STORAGE is just one feature that grants 50G storage.
> FEATURE_50GB_STORAGE_ADD_ON is different: it's 50 accumulated "1GB storage" feature provided by PRODUCT_1GB_SPACE product.
> Thus, these two are not interchangeable in the current form. 

These look very much interchangeable to me. What makes them different? Are we asking the difference between `50 * 1` and `50` or something else? 😕 

add-on as a concept is indeed a different thing than a plan feature (like you seem to agree). what we are trying to do here is create features that are add-ons (so the opposite to what we agree? - mixing things at the conceptual level). I'd think an add-on would just consume a feature to put together a different structure somewhere, not here.

Reply:
> If I understand Christos correctly, the problem right now is that we have two ways of defining storage add ons.

unfortunately, it looks like so.

Reply:
I haven't followed the conversation here too closely, but just chiming in regarding the feature check.

Yes, the existing 50, 100, etc. are the ones used in bundles -- and are individual products mapped to a feature. They have no quantity. 

As the features system isn't aware of quantity, the feature that's mapped to this new product is "Space 1 GB" (signaling that the subscription have this product), and the packages are just "stacks" of 50 or 100 of this product. 

We keep track of quantity on subscription level, but there's no  easy way to do it on feature level (e.g. by mapping a certain quantity to a certain feature), as the feature checks are boolean -- either the subscription has it, or it doesn't.  :)

Reply:
Thanks for chiming in @gmovr . 

I think we are just complicating things. Somehow this FeaturesList doesn't feel like a clearly defined structure. Is an add-on a feature? Does it ever map to the plan features? Would it make sense to return that from an endpoint `/plan-features/[plan_id]`? Are features meant to be something else than what plans or other products are bundled with? 

My guess the answer to all of the above is "no". and if so, then "add-on" has no place in this list. Separate list, fine. Separate hook, even better (and looks like we have one already from the "add-ons" work).

Reply:
I think we are aligned on this. What @southp described and the exploration in #80114 speak of the same concerns to me. 

Maybe we could have defined a separate structure or hook for add-ons here and avoid conflating with the features list. Then a follow-up could be to bridge that structure with the existing `use-add-ons` (which would be a bigger undertaking).

---

Speaking of which - @jeyip - I don't think plan ui package (`plans-features-2023-grid`) should host that hook. As a pointer for #80114 , I would suggest a separate `add-ons` package in the monorepo and export the hook from there. Maybe data-stores, but separate one can also exist to migrate rest of add-ons section there (I wonder when that will be put on pedestal for reuse). Definitely not under plans ui package though IMO. But lets this convo there / #80114

Reply:
Thanks for chiming in y'all!

> Could you help create an issue for this, add it to https://github.com/Automattic/martech/issues/2023, and priortize it as the topmost from here? I'd be grateful :)

Of course -- I'll take care of this on Monday